akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventWriter.scala

```diff
@@ -257,6 +314,98 @@ private[akka] object EventWriter {
+                    persistenceId,
+                    sequenceNumber)
+                  state.copy(waitingForSeqNrLookup = state.waitingForSeqNrLookup :+ ((repr, replyTo)))
+                } else if (fillSequenceNumberGaps && state.latestSeqNr == 0L && sequenceNumber != 1L) {
+                  // we haven't looked up latest yet, and incoming is not the well known 1
+                  askMaxSeqNr(persistenceId, originalErrorDesc = None)
+                  context.log.trace2(
+                    "Seq nr lookup needed for persistence id [{}], adding sequence nr [{}] to pending",
+                    persistenceId,
+                    sequenceNumber)
+                  state.copy(seqNrlookupInProgress = true, waitingForSeqNrLookup = state.waitingForSeqNrLookup :+ ((repr, replyTo)))
+                } else if (sequenceNumber == expectedSeqNr) {
+                  if (state.idle) {
+                    sendToJournal(1, Vector(repr))
+                    state.copy(waitingForReply = Map((repr.sequenceNr, (repr, replyTo))), currentTransactionId = 1)
+                  } else {
+                    // write in progress for pid, add write to batch and perform once current write completes
+                    if (state.waitingForWrite.size >= settings.maxBatchSize) {
+                      replyTo ! StatusReply.error(
+                        s"Max batch reached for pid $persistenceId, at most ${settings.maxBatchSize} writes for " +
+                        "the same pid may be in flight at the same time")
+                      state
+                    } else {
+                      context.log.trace2(
+                        "Writing event in progress for persistence id [{}], adding sequence nr [{}] to batch",
+                        persistenceId,
+                        sequenceNumber)
+                      state.copy(waitingForWrite = state.waitingForWrite :+ ((repr, replyTo)))
+                    }
+                  }
+                } else if (sequenceNumber < expectedSeqNr) {
+                  context.log.trace2("Duplicate seq nr [{}] for persistence id [{}]", sequenceNumber, persistenceId)
+                  replyTo ! StatusReply.success(WriteAck(persistenceId, sequenceNumber)) // duplicate
+                  state
+                } else {
+                  val fillRepr = (expectedSeqNr until sequenceNumber).map { n =>
+                    PersistentRepr(
+                      FilteredPayload,
+                      persistenceId = persistenceId,
+                      sequenceNr = n,
+                      manifest = "", // adapters would be on the producing side, already applied
+                      writerUuid = writerUuid,
+                      sender = akka.actor.ActorRef.noSender)
+                  }.toVector
+                  if (state.idle) {
+                    sendToJournal(1, fillRepr :+ repr)
+                    state.copy(waitingForReply = Map((repr.sequenceNr, (repr, replyTo))), currentTransactionId = 1)
```

Comment:
Hmm, any particular reason for resetting transaction id here, could it maybe be useful to keep accumulating instead so that ids aren't re-used? (otoh we only really use it and never log the value, so maybe not that useful for correlating/bughunting)

Reply:
yeah, I didn't pay attention to the transaction id yet, or I didn't fully understand it. Isn't it reset (before this PR) since the entry in perPidWriteState was removed when all updates for the pid have been completed? Do we want a more global counter? Wasn't the purpose to ignore old replies from the journal after a restart of this actor? but it was something more, since it's per pid.

Reply:
I think it was just to know if we should discard old instance replies yeah.